### PR TITLE
[specific ci=Group1-Docker-Commands] Have fetcher fail quickly on unretryable 4XX errors

### LINKED ISSUE #5966: [Harbor] VCH should not retry pulling when registry returns 4xx
This is another finding along #5950 

When the image doesn't match the project level policy, Harbor will return 412 for the "GET manifest" request.  After receiving 412 VCH keeps retrying 5 times, so it will take around a minute for the "pull" action to really fail.

Generally, I don't think VCH should retry upon receiving a 4xx error code from server.  It probably should fail right away except for receiving a 401.

### PR DESCRIPTION
**Pull request by willsherwood**
Fixes #5966.

Out of the 4XX codes, 'request timeout', 'locked', and 'too many requests' _could_ be resolved with a retry, so I excluded them from the retry block.

### CODE REVIEW

File: pkg/fetcher/fetcher.go [hunk 1]
```diff
@@ -401,6 +406,19 @@ func (u *URLFetcher) IsStatusNotFound() bool {
 	return u.StatusCode == http.StatusNotFound
 }
 
+// IsClientError returns true if status code is an unretryable 4XX error. This includes all
```
Comment (reasonerjt) [inline pkg/fetcher/fetcher.go]: The comment doesn't match the function name?

**Comment by sflxn** [comment]
lgtm.  Will, can you change from Unretryable to Nonretable?

File: pkg/fetcher/fetcher.go [hunk 2]
```diff
@@ -401,6 +406,19 @@ func (u *URLFetcher) IsStatusNotFound() bool {
...
 
+// IsNonretryableClientError returns true if status code is a nonretryable 4XX error. This includes
+// all 4XX errors except 'request timeout', 'locked', and 'too many requests'. Additionally, it
+// does not return 'not found' or 'unauthorized' errors since those are handled uniquely.
+func (u *URLFetcher) IsNonretryableClientError() bool {
+	s := u.StatusCode
+	return 400 <= s && s < 500 &&
+		s != http.StatusNotFound &&
+		s != http.StatusUnauthorized &&
+		s != http.StatusRequestTimeout &&
```
Comment (caglar10ur) [inline pkg/fetcher/fetcher.go]: Why we are not retrying in StatusRequestTimeout case?

File: pkg/fetcher/fetcher.go [hunk 3]
```diff
@@ -401,6 +406,19 @@ func (u *URLFetcher) IsStatusNotFound() bool {
 	return u.StatusCode == http.StatusNotFound
 }
 
+// IsNonretryableClientError returns true if status code is a nonretryable 4XX error. This includes
+// all 4XX errors except 'request timeout', 'locked', and 'too many requests'. Additionally, it
+// does not return 'not found' or 'unauthorized' errors since those are handled uniquely.
+func (u *URLFetcher) IsNonretryableClientError() bool {
```
Comment (caglar10ur) [inline pkg/fetcher/fetcher.go]: Would it be more readable/less error prone/and easy to modify if we just create a map of errors to bools and do a lookup here?

Comment (chengwang86) [inline pkg/fetcher/fetcher.go]: One more comment:

Maybe we can do something like this:
```
if u.IsNonretryableClientError() {
    if u.IsStatusUnauthorized() {
     }
     if u.IsStatusNotFound() {
    }
    ....
    err = fmt.Errorf("Nonretryable error '%s (%d)': URL %s.", http.StatusText(u.StatusCode), u.StatusCode, url)
    return nil, nil, DoNotRetry{Err: err}
}
```
By doing this, you don't need to exclude `NotFound` and `unauthoized` from `IsNonretryableClientError`.

File: pkg/fetcher/fetcher.go (hunk 2 above)
Comment (willsherwood) [inline pkg/fetcher/fetcher.go]: Ah, I misread the description of status 408. StatusRequestTimeout should retry upon failure.

**Pull request merged**